### ISSUE TITLE
[BUG] installer: DHCP test results in multiple leases for the same MAC address and breaks DDNS updates

### ISSUE THREAD

**Issue by tserong** [harvester/harvester#5926]
**Describe the bug**
This one is pretty obscure, but looks related to https://github.com/harvester/harvester/issues/3428. I ran into it while working on https://github.com/harvester/harvester/issues/1444. It only happens if your DHCP server is configured to hand out addresses from a pool (rather than using fixed IP addresses per MAC address), and you're also doing dynamic DNS updates.

I have the ISC DHCP server running on my local network, configured to hand out addresses within the range 192.168.4.100-199, and to do dynamic DNS updates for hosts within that range. I was testing installing Harvester on a single node on that network, and had added that node to my DHCP server's `/etc/dhcpd.conf` as follows:

```
  host o990 {
    ddns-hostname o990;
    hardware ethernet 18:03:73:be:f1:f4;
  }
```

When that host boots, it should be given an IP address, and we should get a dynamic DNS update mapping that IP address back to `o990.example.com`. OK, my local network is not called "example.com", but you get the idea, right?

Here's what I get from my DHCP server logs (`journalctl -f -u named -u dhcpd`) when running through the harvester installer. This first snippet is from just after applying the network config:

```
May 30 22:03:16 seshat dhcpd[8048]: DHCPDISCOVER from 18:03:73:be:f1:f4 via eth0
May 30 22:03:17 seshat dhcpd[8048]: DHCPOFFER on 192.168.4.197 to 18:03:73:be:f1:f4 (baremetal) via eth0
May 30 22:03:17 seshat dhcpd[8048]: DHCPREQUEST for 192.168.4.197 (192.168.4.4) from 18:03:73:be:f1:f4 (baremetal) via eth0
May 30 22:03:17 seshat dhcpd[8048]: DHCPACK on 192.168.4.197 to 18:03:73:be:f1:f4 (baremetal) via eth0
May 30 22:03:19 seshat dhcpd[8048]: DHCPDISCOVER from 18:03:73:be:f1:f4 via eth0
May 30 22:03:20 seshat dhcpd[8048]: DHCPOFFER on 192.168.4.187 to 18:03:73:be:f1:f4 via eth0
May 30 22:03:20 seshat dhcpd[8048]: DHCPREQUEST for 192.168.4.187 (192.168.4.4) from 18:03:73:be:f1:f4 via eth0
May 30 22:03:20 seshat dhcpd[8048]: DHCPACK on 192.168.4.187 to 18:03:73:be:f1:f4 via eth0
May 30 22:03:20 seshat named[7909]: client @0x7fd094012730 192.168.4.4#25213: updating zone 'example.com/IN': adding an RR at 'o990.example.com' A 192.168.4.187
May 30 22:03:20 seshat named[7909]: client @0x7fd094012730 192.168.4.4#25213: updating zone 'example.com/IN': adding an RR at 'o990.example.com' TXT "000af9802f96a58a4ebfe48289939367c8"
May 30 22:03:20 seshat dhcpd[8048]: Added new forward map from o990.example.com to 192.168.4.187
```

Here's what I get a bit later when I configure the VIP via DHCP:

```
May 30 22:03:51 seshat dhcpd[8048]: DHCPDISCOVER from 1e:72:69:87:9c:15 via eth0
May 30 22:03:52 seshat dhcpd[8048]: DHCPOFFER on 192.168.4.188 to 1e:72:69:87:9c:15 via eth0
May 30 22:03:52 seshat dhcpd[8048]: DHCPREQUEST for 192.168.4.188 (192.168.4.4) from 1e:72:69:87:9c:15 via eth0
May 30 22:03:52 seshat dhcpd[8048]: DHCPACK on 192.168.4.188 to 1e:72:69:87:9c:15 via eth0
```

The VIP is fine, but note from the first log snippet that the same MAC address (`18:03:73:be:f1:f4`) has been given two different IP addresses (`192.168.4.197` and `192.168.4.187`). The first IP address is the one that results from applying the network configuration. The second IP address is the one we get when the harvester installer calls `getIPThroughDHCP()`, but it's the _second_ IP address (the test one) that gets a record added to DNS!

If I check `/var/lib/dhcp/db/dhcpd.leases`, I can see two separate leases for the same MAC address:

```
lease 192.168.4.187 {
  starts 4 2024/05/30 12:03:20;
  ends 4 2024/06/06 12:03:20;
  cltt 4 2024/05/30 12:03:20;
  binding state active;
  next binding state free;
  rewind binding state free;
  hardware ethernet 18:03:73:be:f1:f4;
  set ddns-txt = "000af9802f96a58a4ebfe48289939367c8";
  set ddns-fwd-name = "o990.example.com";
}
lease 192.168.4.197 {
  starts 4 2024/05/30 12:19:12;
  ends 4 2024/06/06 12:19:12;
  cltt 4 2024/05/30 12:19:12;
  binding state active;
  next binding state free;
  rewind binding state free;
  hardware ethernet 18:03:73:be:f1:f4;
  uid "\377s\276\361\364\000\001\000\001-\353'|\030\003s\276\361\364";
  client-hostname "baremetal";
}
```

Later, when booting into the installed system, I see this:

```
May 30 22:19:12 seshat dhcpd[8048]: DHCPREQUEST for 192.168.4.197 from 18:03:73:be:f1:f4 (baremetal) via eth0
May 30 22:19:12 seshat dhcpd[8048]: DHCPACK on 192.168.4.197 to 18:03:73:be:f1:f4 (baremetal) via eth0
May 30 22:19:12 seshat named[7909]: client @0x7fd0940044e0 192.168.4.4#25213: updating zone 'example.com/IN': update unsuccessful: o990.example.com: 'name not in use' prerequisite not satisfied (YXDOMAIN)
May 30 22:19:12 seshat named[7909]: client @0x7fd0940044e0 192.168.4.4#25213: updating zone 'example.com/IN': update unsuccessful: o990.example.com/TXT: 'RRset exists (value dependent)' prerequisite not satisfied (NXRRSET)
May 30 22:19:12 seshat dhcpd[8048]: Forward map from o990.example.com to 192.168.4.197 FAILED: Has an address record but no DHCID, not mine.
```

So my host gets the correct lease (`192.168.4.197`, i.e. the one it got when the network config was applied during install) on reboot, but the DNS has been trashed due to that second test DHCP request for the same MAC address. I _want_ `o990.example.com` to point to `192.168.4.197`, but instead it gives me `192.168.4.187`.

There must be some subtle difference between the DHCP request made when restarting the network (which I note includes the client hostname), and the test DHCP request made via the golang client, which confuses the DHCP sever and results in two leases. I suspect that whatever this difference is, is also the cause of the Windows DHCP server failing in https://github.com/harvester/harvester/issues/3428. Essentially, in both cases, we're just freaking out the DHCP server.

Here's what I get in the server logs if I remove the call to `getIPThroughDHCP()` (as in https://github.com/bk201/harvester-installer/commit/7d0cbf546f34f2225ac4c316a68e71d92d9379ed):

```
[ applying network config ]
May 31 21:20:46 seshat dhcpd[8739]: DHCPDISCOVER from 18:03:73:be:f1:f4 via eth0
May 31 21:20:47 seshat dhcpd[8739]: DHCPOFFER on 192.168.4.197 to 18:03:73:be:f1:f4 (rancher) via eth0
May 31 21:20:47 seshat dhcpd[8739]: DHCPREQUEST for 192.168.4.197 (192.168.4.4) from 18:03:73:be:f1:f4 (rancher) via eth0
May 31 21:20:47 seshat dhcpd[8739]: DHCPACK on 192.168.4.197 to 18:03:73:be:f1:f4 (rancher) via eth0

[ requesting VIP ]
May 31 21:20:55 seshat dhcpd[8739]: DHCPDISCOVER from 22:47:4d:9e:ba:c3 via eth0
May 31 21:20:56 seshat dhcpd[8739]: DHCPOFFER on 192.168.4.187 to 22:47:4d:9e:ba:c3 via eth0
May 31 21:20:56 seshat dhcpd[8739]: DHCPREQUEST for 192.168.4.187 (192.168.4.4) from 22:47:4d:9e:ba:c3 via eth0
May 31 21:20:56 seshat dhcpd[8739]: DHCPACK on 192.168.4.187 to 22:47:4d:9e:ba:c3 via eth0

[ first boot of installed system ]
May 31 21:36:13 seshat dhcpd[8739]: Wrote 0 deleted host decls to leases file.
May 31 21:36:13 seshat dhcpd[8739]: Wrote 0 new dynamic host decls to leases file.
May 31 21:36:13 seshat dhcpd[8739]: Wrote 40 leases to leases file.
May 31 21:36:13 seshat dhcpd[8739]: DHCPREQUEST for 192.168.4.197 from 18:03:73:be:f1:f4 (rancher) via eth0
May 31 21:36:13 seshat dhcpd[8739]: DHCPACK on 192.168.4.197 to 18:03:73:be:f1:f4 (o990) via eth0
May 31 21:36:13 seshat named[8771]: client @0x7f03fc047ef0 192.168.4.4#43416: updating zone 'example.com/IN': adding an RR at 'o990.example.com' A 192.168.4.197
May 31 21:36:13 seshat named[8771]: client @0x7f03fc047ef0 192.168.4.4#43416: updating zone 'example.com/IN': adding an RR at 'o990.example.com' TXT "3165eb21fbbe243c489bf4a4a638808baf"
May 31 21:36:13 seshat dhcpd[8739]: Added new forward map from o990.example.com to 192.168.4.197
```

That's fine. `o990.example.com` gets mapped back to `192.168.4.197`, which is the IP address of that host.

TL;DR: using `getIPThroughDHCP()` to verify DHCP is working causes weird problems in certain circumstances.

**To Reproduce**
See above. If necessary I can probably figure out a config we could use with ipxe-examples to make a reproducer. For now though I just wanted to document this cursed thing ;-)

**Expected behavior**
DHCP server does not freak out. DDNS works properly.

**Environment**
 - Harvester ISO version: v1.3, but presumably all versions.
 - Underlying Infrastructure (e.g. Baremetal with Dell PowerEdge R630): baremetal on an old dell optiplex desktop system

**Additional context**
My thinking on this so far is that we should just drop the call to `getIPThroughDHCP()`. I found it causes this weird DDNS problem, and we know from https://github.com/harvester/harvester/issues/3428 that it can cause trouble with Windows DHCP servers. If we really do need to do an extra test to make sure DHCP works, maybe we can:
1. Try a request much like we do for the VIP, from a different/virtual MAC address (this would avoid duplicate leases), or, ...
2. Do it by inference, i.e. check the network config before activating DHCP when applying network settings, then after the network settings are applied, verify that the system now has an IP address that it didn't have before.
